tests/utils/test_http.py

```diff
@@ -0,0 +1,57 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from warehouse.utils.http import is_safe_url
+
+
+class TestHttp:
+
+    # (MOSTLY) FROM https://github.com/django/django/blob/
+    # 011a54315e46acdf288003566b8570440f5ac985/tests/utils_tests/test_http.py
+    def test_is_safe_url(self):
+        # A None URL is not safe!
+        assert not is_safe_url(None)
+
+        for bad_url in ('http://example.com',
```

Comment:
This should use pytest parameterized tests, like:

```python

class TestIsSafeURL:

    @pytest.mark.parametrize(
        "url",
        [
            None,
            "http://example.com",
            "http:///example.com",
            ...,
        ],
    )
    def test_rejects_bad_url(self, url):
        assert not is_safe_url(url, host="testserver")

    @pytest.mark.parametrize(
        "url",
        [
            "/view/?param=http://example.com",
            "/view/?param=https://example.com",
            ...,
        ],
    )
    def test_accepts_good_url(self, url):
        assert is_safe_url(url, host="testserver")
```

This will make the tests themselves a lot more declarative and will make it so a failure on a single value early on won't cause the rest of the tests to not run.

Additionally you don't need to provide an explicit assertion message, pytest will provide a rich repr for the assertion failure that gives more details than a standard assert error.